fdbserver/storageserver.actor.cpp

```diff
@@ -2470,6 +2509,440 @@ ACTOR Future<Void> getKeyValuesQ(StorageServer* data, GetKeyValuesRequest req)
+		if (type == Tuple::BYTES || type == Tuple::UTF8) {
+			std::string s = hopKeyFormatTuple.getString(i).toString();
+			auto sz = s.size();
+
+			// Handle escape.
+			bool escaped = false;
+			size_t p = 0;
+			while (true) {
+				size_t found = s.find("{{", p);
+				if (found == std::string::npos) {
+					break;
+				}
+				s.replace(found, 2, "{");
+				p += 1;
+				escaped = true;
+			}
+			p = 0;
+			while (true) {
+				size_t found = s.find("}}", p);
+				if (found == std::string::npos) {
+					break;
+				}
+				s.replace(found, 2, "}");
+				p += 1;
+				escaped = true;
+			}
+			if (escaped) {
+				// If the element uses escape, cope the escaped version.
+				hopKeyTuple.append(s);
+			}
+			// {K[??]} or {V[??]}
+			else if (sz > 5 && s[0] == '{' && (s[1] == 'K' || s[1] == 'V') && s[2] == '[' && s[sz - 2] == ']' &&
+			         s[sz - 1] == '}') {
+				int idx;
+				try {
+					idx = std::stoi(s.substr(3, sz - 5));
+				} catch (std::exception& e) {
+					throw hop_bad_index();
+				}
+				Tuple* referenceTuple;
+				if (s[1] == 'K') {
+					// Use keyTuple as reference.
+					if (!keyTuple.present()) {
+						// May throw exception if the key is not parsable as a tuple.
+						keyTuple = Tuple::unpack(keyValue->key);
+					}
+					referenceTuple = &keyTuple.get();
+				} else if (s[1] == 'V') {
+					// Use valueTuple as reference.
+					if (!valueTuple.present()) {
+						// May throw exception if the value is not parsable as a tuple.
+						valueTuple = Tuple::unpack(keyValue->value);
+					}
+					referenceTuple = &valueTuple.get();
+				} else {
+					ASSERT(false);
+					throw internal_error();
+				}
+
+				if (idx < 0 || idx >= referenceTuple->size()) {
+					throw hop_bad_index();
+				}
+				hopKeyTuple.append(referenceTuple->subTuple(idx, idx + 1));
+			} else if (s == "{...}") {
+				// Range query.
+				if (i != hopKeyFormatTuple.size() - 1) {
+					// It must be the last element of the hop info tuple
+					throw hop_bad_range_decriptor();
+				}
+				// Every record will try to set it. It's ugly, but not wrong.
+				isRangeQuery = true;
+				// Do not add it to the hop key.
+			} else {
+				// If the element is a string but neither escaped nor descriptors, just copy it.
+				hopKeyTuple.append(hopKeyFormatTuple.subTuple(i, i + 1));
+			}
+		} else {
+			// If the element not a string, just copy it.
+			hopKeyTuple.append(hopKeyFormatTuple.subTuple(i, i + 1));
+		}
+	}
+	return hopKeyTuple.getDataAsStandalone();
+}
+
+TEST_CASE("/fdbserver/storageserver/constructHopKey") {
+	Key key = Tuple().append("key-0"_sr).append("key-1"_sr).append("key-2"_sr).getDataAsStandalone();
+	Value value = Tuple().append("value-0"_sr).append("value-1"_sr).append("value-2"_sr).getDataAsStandalone();
+	state KeyValueRef kvr(key, value);
+	{
+		Tuple hopInfoTuple = Tuple()
+		                         .append("normal"_sr)
+		                         .append("{{escaped}}"_sr)
+		                         .append("{K[2]}"_sr)
```

Comment:
This K[], V[], is a little hard to read, still don't know how is the hopInfo is constucted and used.